electrum/plugins/bitbox02/bitbox02.py

```diff
@@ -0,0 +1,401 @@
+            return False
+        return True
+
+    def show_address(self, bip32_path, address_type):
+        print("getting address on: ", address_type, bip32_path)
+        address_keypath, is_mainnet = self.bip32_list_from_keypath_string(bip32_path)
+        if address_type == "p2wpkh" or address_type == "p2pkh" or address_type == "standard":
+            if is_mainnet:
+                coin_network = bitbox02.btc.BTC
+                script = bitbox02.btc.SCRIPT_P2WPKH
+            else:
+                coin_network = bitbox02.btc.TBTC
+                script = bitbox02.btc.SCRIPT_P2WPKH
+        elif address_type == "p2wpkh-p2sh":
+            if is_mainnet:
+                coin_network = bitbox02.btc.BTC
+                script = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+            else:
+                coin_network = bitbox02.btc.TBTC
+                script = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+        else:
+            raise Exception("invalid keypath address_type")
+
+        address = self.device.btc_pub(
+                keypath = address_keypath,
+                output_type = bitbox02.btc.BTCPubRequest.ADDRESS,  # pylint: disable=no-member
+                script_type = script,
+                coin = coin_network,
+            )
+        return address
+
+    def sign_transaction(self, keystore, tx):
+        if tx.is_complete():
+            return
+
+        inputhasharray = []
+        hasharray = []
+        pubkeyarray = []
+        network = bitbox02.btc.BTC
+        if constants.net.TESTNET:
+            network = bitbox02.btc.TBTC
+
+        tx_script_type = bitbox02.btc.SCRIPT_P2WPKH
+
+        # Build BTCInputType list
+        inputs = []
+        for txin in tx.inputs():
+            if txin.is_coinbase():
+                self.give_error("Coinbase transactions are not supported")
+            _, full_path = keystore.find_my_pubkey_in_txinout(txin)
+
+            inputs.append({
+                "prev_out_hash": txin.prevout.txid,
+                "prev_out_index": txin.prevout.out_idx,
+                "prev_out_value": txin.value_sats(),
+                "sequence": 0xffffffff,
+                "keypath": full_path,
+            })
+
+            if full_path[0] == 49 + HARDENED:
+                tx_script_type = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+
+        # Build BTCOutputType list
+        outputs = []
+        for txout in tx.outputs():
+            assert txout.address
+            # check for change
+            if txout.is_change:
+                _, full_path = keystore.find_my_pubkey_in_txinout(txout)
+
+                outputs.append(
+                    bitbox02.BTCOutputInternal(
+                        keypath = full_path,
+                        value = txout.value,
+                    )
+                )
+            else:
+                addrtype, pubkey_hash = bitcoin.address_to_hash(txout.address)
+                if addrtype == constants.net.ADDRTYPE_P2PKH:
```

Comment:
you are mixing up result types. `addrtype` is the segwit version in case of a segwit addr, which is `0`, so you might be running into the `ADDRTYPE_P2PKH` case,

best to make your own result addrtype enum in `address_to_hash` with the three possible values.

Reply:
I am using bitcoin's WIF_SCRIPT_TYPES dict now to transport the types. I noticed that we do not specify outputs for `p2wsh-sh` and `p2wpkh-p2sh`, do we need to include these as well?

Reply:
This is the wrong type too, as those are input types, not output types ;) The supported output types are P2SH, P2WPKH and P2WSH. 

Addresses don't contain a redeem script, so the `-p2sh` is not meaningful there. (It is meaningful in change addresses to derive the address, but that script type is already provided in the `BTCSignInitRequest`).

We have a bit of a special thing in the bb02 that we explicitly transmit the type of the address and the data to derive the address. The alternative would have been to take the raw address and parse the relevant data from it. In any case, this means the parsing happens in software (like here) and you just need to provide the pubkey or script hash plus the type (one of the three above). I still think you can just use the btc.proto's `BTCOutputType`, it will be the most appropriate and correct type, as this is what `BTCOutputExternal` actually wants.